src/D2L.Bmx/PrintHandler.cs

```diff
@@ -12,54 +17,56 @@ internal class PrintHandler {
 		string? output
 	) {
 
-		// TODO: Get values of org, user, account, and role from the config file and assign them locally. This way we can check if
-		// values are null to see if they have been set through the commandline or config file
+		var config = _configProvider.GetConfiguration();
 
 		// ask user to input org if org flag isn't set
 		if( string.IsNullOrEmpty( org ) ) {
-			// TODO: also print out --help output
-			Console.WriteLine( "Error: required flag(s) \"org\"" );
-			return;
+			if( !string.IsNullOrEmpty( config.Org ) ) {
+				org = config.Org;
+			} else {
+				org = ConsolePrompter.PromptOrg();
+			}
 		};
 
 		// ask user to input username if user flag isn't set
 		if( string.IsNullOrEmpty( user ) ) {
-			Console.Write( "Okta Username: " );
-			// there is currently no input validation for the username in the golang build, but it can be added
-			user = Console.ReadLine();
+			if( !string.IsNullOrEmpty( config.User ) ) {
+				user = config.User;
+			} else {
+				user = ConsolePrompter.PromptUser();
+			}
 		};
 
 		// Asks for user password input, or logs them in through caches
 		Authenticator.Authenticate( org, user, nomask );
 
-		// TODO: replace placeholder values with actual values
+		// TODO: replace placeholder values with actual values. Get accounts and roles list from AWS and pass it into the prompter
 		var accounts = new[] { "Dev-Slims", "Dev-Toolmon", "Int-Dev-NDE" };
 		var roles = new[] { "Dev-Slims-ReadOnly", "Dev-Slims-Admin" };
 
 		if( string.IsNullOrEmpty( account ) ) {
-			Console.WriteLine( "Available accounts:" );
-			for( int i = 0; i < accounts.Length; i++ ) {
-				Console.WriteLine( $"[{i + 1}] {accounts[i]}" );
+			if( !string.IsNullOrEmpty( config.Account ) ) {
+				account = config.Account;
+			} else {
+				account = ConsolePrompter.PromptAccount( accounts );
 			}
-			Console.Write( "Select an account: " );
-			if( !int.TryParse( Console.ReadLine(), out int index ) || index > accounts.Length || index < 1 ) {
-				Console.WriteLine( "Error: Invalid selection" );
-				return;
-			}
-			account = accounts[index - 1];
 		}
 
 		if( string.IsNullOrEmpty( role ) ) {
-			Console.WriteLine( "Available roles:" );
-			for( int i = 0; i < roles.Length; i++ ) {
-				Console.WriteLine( $"[{i + 1}] {roles[i]}" );
+			if( !string.IsNullOrEmpty( config.Role ) ) {
+				role = config.Role;
+			} else {
+				role = ConsolePrompter.PromptRole( roles );
 			}
-			Console.Write( "Select a role: " );
-			if( !int.TryParse( Console.ReadLine(), out int index ) || index > roles.Length || index < 1 ) {
-				Console.WriteLine( "Error: Invalid selection" );
-				return;
+		}
+
+		// if duration equals zero, then the duration flag is not set. Users cannot set duration to 0 because of flag validation
+		if( duration == 0 ) {
```

Comment:
Wait... the default value for the duration flag is 0 not null? Our signature `int? duration` is a lie then...

Right... this is a consequence of the ambiguity of `T?` in generics. I'd prefer we make `durationOption` an `Option<int?>` then (not `Option<int>`), then we'd do a null check instead of a zero check here, and the code should be clearer & we won't need the comment.

Reply:
More context:
* In generic code, `T` and `T?` mean the same thing for value types (e.g. `int`) (i.e. the question mark is meaningless), but they do have different meanings for reference types (e.g. `string`).
* This is in part because reference type nullability is only a compile time construct, not a runtime constraint - `string` and `string?` are the same type. On the other hand, value types can never be null, so `int?` is actually an entirely different runtime type [`Nullable<int>`](https://learn.microsoft.com/en-us/dotnet/api/system.nullable-1?view=net-7.0)
* The signature of [`GetValueForOption`](https://learn.microsoft.com/en-us/dotnet/api/system.commandline.parsing.parseresult.getvalueforoption?view=system-commandline) says it should return an object of type `T` and it _may_ be null, which works intuitively for reference types, but for value types, because they can never be null, "may be null" is meaningless. It doesn't return `Nullable<int>` for `Option<int>`, because that's a different type from the type parameter `int`. If we pass it an `Option<int?>` though, it should return `int?` i.e. `Nullable<int>` as expected.